Add Potion Types

sschr15 opened the pull request:
There were a couple issues with the language files including a name without a translation and two definitions for one item

sschr15 committed:
Fix issues with language files

Tmtravlr commented on src/main/resources/assets/additions/lang/en_us.lang:
```diff
@@ -23,7 +23,6 @@ commands.additions.throw.success=%1$s successfully threw %2$s
 commands.additions.move.usage=/additions.move <global | relative | facing | toward>
 commands.additions.move.global.usage=/additions.move global <x speed> <y speed> <z speed>
 commands.additions.move.facing.usage=/additions.move relative <horizontal speed forward/back> [speed y] [horizontal speed right/left]
-commands.additions.move.facing.usage=/additions.move facing <speed forward/back> [speed up/down] [speed right/left]
```
Oof yeah... That's a copypasta error. This line's fine actually. The line above it should be commands.additions.move.relative.usage

Looks like I'm never actually using the relative usage in the command though, I'll have to fix that too.

Tmtravlr commented on src/main/resources/assets/additions/lang/zh_cn.lang:
```diff
@@ -23,7 +23,6 @@ commands.additions.throw.success=%1 $s 成功投掷 %2 $s
 commands.additions.move.usage=/additions.move <global | relative | facing | toward>
 commands.additions.move.global.usage=/additions.move global <x 速度> <y 速度> <z 速度>
 commands.additions.move.facing.usage=/additions.move relative <horizontal speed forward/back> [speed y] [horizontal speed right/left]
-commands.additions.move.facing.usage=/additions.move facing <speed forward/back> [speed up/down] [speed right/left]
```
Same here

Tmtravlr commented on src/main/resources/assets/additions/lang/en_us.lang:
```diff
@@ -11,7 +11,7 @@ commands.additions.damage.success=Applied %1$s hitpoints of %2$s damage to %3$s
 commands.additions.damageitem.usage=/additions.damageitem <entity|block>
 commands.additions.damageitem.block.usage=/additions.damageitem entity <entity> <slot> <amount>
-commands.additions.damageitem.entity.usage/additions.damageitem block <block> <x> <y> <z> <slot> <amount>
+commands.additions.damageitem.entity.usage=/additions.damageitem block <block> <x> <y> <z> <slot> <amount>
```
Nice catch!

Tmtravlr reviewed (changes requested):
Thanks for the review! 

Just need to change the first commands.additions.move.facing.usage to commands.additions.move.relative.usage instead of removing the second one

sschr15 committed:
Create PotionTypeAdded and AdditionTypePotionType
Also added some TODO with other things to do eventually

sschr15 committed:
Potion Types now work. Tomorrow is color-adding day.

sschr15 committed:
Nope no color adding, just there for future construction.

Made it work with hopefully few crashes, please report any that you find!

(also cleaned up a little)

sschr15 committed:
Re-fixed language files

sschr15 committed:
made postfix consistent

and intellij optimized imports

Tmtravlr commented on src/main/resources/assets/additions/lang/zh_cn.lang:
```diff
@@ -1153,3 +1153,26 @@ gui.edit.texture.block.facing.top.label=顶部/底部材质
 
 gui.edit.texture.block.pane.face.message=为%s的面部选择一个材质.
 gui.edit.texture.block.pane.face.label=面部材质
+
+## NEEDS TRANSLATION
+gui.view.addon.potionTypes.label=View Potion Types Added (%s)
+gui.view.addon.potionTypes.title=Potion Types for %s
+gui.view.addon.potionTypes.none.message=There are no potion types for %s
+gui.edit.potionType.title=Create New Potion Type
+gui.edit.potionType.id.label=Potion Type Id
+gui.edit.potionType.baseName.label=Normal Potion Name
+gui.edit.potionType.effects.label=Potion Effects
+gui.edit.potionType.inputInformation=All of the options below are optional and are automatically generated if not stated.
+gui.edit.potionType.splashName.label=Splash Potion Name
+gui.edit.potionType.lingeringName.label=Lingering Potion Name
+gui.edit.potionType.arrowName.label=Tipped Arrow Name
+gui.edit.potionType.arrowName.info=A tipped arrow is automatically created if this potion type has at least one effect.
+gui.edit.potionType.potionColor.label=Potion Color
+gui.edit.potionType.potionColor.switch.label=Use Custom Potion Color
+gui.edit.potionType.problem.title=Problem creating potion type
+gui.edit.potionType.problem.noBase=A Potion Type requires a base name!
+gui.edit.potionType.success.title=Updated Potion Type
+gui.edit.potionType.success.message=Successfully updated the potion type.
+gui.edit.potionType.restart.message=Potion type successfully created! You must restart the game for it to show up ingame and in some lists. Restart now? (Will close your game)
+gui.edit.potionType.restartToRegister.first=Restart the game to finish
+gui.edit.potionType.restartToRegister.second=adding this potion type.
```
Don't add these lines here, I'll send the changes to LiuMang, and he will add them with the translations

sschr15 replied:
Committed, removed

Tmtravlr commented on src/main/java/com/tmtravlr/additions/type/AdditionTypeItem.java:
```diff
@@ -153,7 +151,7 @@ public void saveAddition(Addon addon, IItemAdded addition) {
 		
 		File itemFolder = new File(addon.addonFolder, FOLDER_NAME);
-		
+		 //TODO become consistent
```
What is this for?

sschr15 committed:
Remove TODOs

sschr15 committed:
Clean up Gson building

sschr15 committed:
Add Tmtravlr's commit to my branch real quick (#1)

Fixed the move command so it shows the relative usage (not just facing)

sschr15 commented:
There's currently a problem with my Gson-related code, don't merge yet!

sschr15 committed:
Fixed other Gson-related problem to work as intended :)

sschr15 committed:
Merge remote-tracking branch 'fork/fork/dev' into dev

sschr15 commented:
Alright, problem fixed. We're good to go!

sschr15 committed:
Added comments for why things are the way they are

sschr15 committed:
bump version to beta version of 0.5.3

Tmtravlr commented on gradle.properties:
```diff
@@ -2,7 +2,7 @@
 org.gradle.jvmargs=-Xmx3G
 
-version=0.5.2
+version=0.5.3-b1
```
What's the b1 for?

sschr15 replied:
b for build (pre-release thing), 1 because it's the first build

sschr15 replied:
the first build of the next version, should I change it?

Tmtravlr commented on src/main/java/com/tmtravlr/additions/ConfigLoader.java:
```diff
@@ -39,6 +40,7 @@ public static void loadConfigFile(FMLPreInitializationEvent event) {
 		
 		skipReloadingResources = config.get("editor", "Skip Reloading Resources", false, "If true, skips any resource reloading. Good if you want to add/edit a bunch of things quickly,\nwithout interruption. You can still reload the resources by restarting the game or by hitting F3+T ingame.");
 		renderItemsInLists = config.get("editor", "Render Items in Item Selectors?", true, "If false, won't render items in item/block/ore dictionary selectors. Set to false if you are\ngetting crashes when looking at those selectors.");
+		prettyPrintGeneratedFiles = config.get("editor", "Pretty-Print Generated Json Files", true, "If false, will create minimized Json files rather than easily readable files.");
```
I think it would make sense for this to default to false (so people can change it to true if they want readable json files, but the addons are smaller otherwise)

Tmtravlr commented on src/main/java/com/tmtravlr/additions/gui/view/components/input/GuiComponentIntegerInput.java:
```diff
@@ -25,11 +23,11 @@
 
 	public GuiEdit editScreen;
-	public Predicate<String> validator;
+	public java.util.function.Predicate<String> validator;
```
Whoa, the import above has the wrong Predicate class somehow... it should have this one.

sschr15 replied:
It seems like other classes are built the same way... attempting to fix now

Tmtravlr commented on src/main/java/com/tmtravlr/additions/gui/view/components/input/GuiComponentIntegerInput.java:
```diff
@@ -202,12 +193,23 @@ public void setFocused(boolean focused) {
-			return Integer.parseInt(this.getText());
+			int theInt = Integer.parseInt(this.getText());
+			lastException = null;
+			return theInt;
```
Suggestion: Instead of checking the last exception, check that this.getText() is not empty (so it doesn't even need to throw an exception)
```suggestion
			return this.getText().isEmpty() ? Math.max(0, this.minimum) : Integer.parseInt(this.getText());
```

Tmtravlr commented on src/main/java/com/tmtravlr/additions/gui/view/edit/GuiEditPotionType.java:
```diff
@@ -0,0 +1,195 @@
+
+import javax.annotation.Nullable;
+import java.util.List;
+
```
Forgot the javadoc! =)

Tmtravlr commented on src/main/java/com/tmtravlr/additions/gui/view/edit/GuiEditPotionType.java:
```diff
@@ -0,0 +1,195 @@
+
+        // These two components aren't added to the screen by default as potion types don't usually have a way to list the color
+        this.useUniqueColorInput = new GuiComponentBooleanInput(I18n.format("gui.edit.potionType.potionColor.switch.label"), this);
+        // Reason for noinspection: readability. The suggested change was to `!this.isNew && this.addition.hasColor`
```
I think the suggestion is pretty readable 😛

Tmtravlr commented on src/main/java/com/tmtravlr/additions/type/AdditionTypeItemMaterial.java:
```diff
@@ -49,12 +43,11 @@
 	public static final AdditionTypeItemMaterial INSTANCE = new AdditionTypeItemMaterial();
 	
-	private static final Gson GSON = new GsonBuilder()
+		private static final Gson GSON = GeneralUtils.newBuilder()
```
An extra tab got in here somehow

Tmtravlr commented on src/main/java/com/tmtravlr/additions/util/JsonGenerator.java:
```diff
@@ -13,6 +13,8 @@
  */
 public class JsonGenerator {
 
+	//TODO add a GsonBuilder to auto-decide pretty printing
```
Can remove this todo now

Tmtravlr commented on src/main/java/com/tmtravlr/additions/addon/potiontypes/PotionTypeAdded.java:
```diff
@@ -0,0 +1,148 @@
+    }
+
+    public PotionTypeAdded setColor(int color) {
+        // if color is -1, the color will be removed
+        if (color == -1) this.color = -1;
```
Hmm... you could make the color an Integer instead of an int, so then null will be the not set value instead of -1. 

Actually it looks like the line below will set it to 0 if it is -1 anyway...

Tmtravlr commented on src/main/java/com/tmtravlr/additions/gui/type/button/GuiAdditionTypeButtonPotionType.java:
```diff
@@ -0,0 +1,44 @@
+    // please ignore my inconsistencies with Tmtravlr's code: I like capital hex numbers
+    public static final int BUTTON_COLOR_HOVER = 0xFF3C84E2;
+    public static final int BUTTON_COLOR_LIGHT = 0xFF6FB7FF;
+    public static final int BUTTON_COLOR_DARK  = 0xFF0951AF;
```
I love the pretty colors, but they are noticeably lighter then the rest of the buttons. Maybe they should be darkened just a tad.

Tmtravlr commented on src/main/java/com/tmtravlr/additions/gui/type/card/GuiAdditionCardPotionType.java:
```diff
@@ -0,0 +1,108 @@
+    protected void drawCardInfo(int mouseX, int mouseY) {
+        int itemDisplayTop = this.y + this.height / 2 - 10;
+        int columnWidth = this.getColumnWidth();
+        ItemStack itemStack;
+        //noinspection ConstantConditions
+        if (PotionType.REGISTRY.getNameForObject(this.addition) == null) {
+            itemStack = new ItemStack(Items.POTIONITEM);
+            itemStack.setTranslatableName(this.addition.getBaseName());
+            NBTTagCompound stackTag = itemStack.getTagCompound();
+            NBTTagCompound display = stackTag.getCompoundTag("display");
+            NBTTagList lore = display.getTagList("Lore", 8);
+            lore.appendTag(new NBTTagString(I18n.format("gui.edit.potionType.restartToRegister.first")));
+            lore.appendTag(new NBTTagString(I18n.format("gui.edit.potionType.restartToRegister.second")));
+            display.setTag("Lore", lore);
+            stackTag.setTag("display", display);
+            itemStack.setTagCompound(stackTag);
+        }
+        else itemStack = this.addition.getPotionItemStack();
```
That's a lotta code to get the item stack each rendering frame... Maybe the item stack should be created just once in the constructor method and saved to an instance variable with the text strings.

Tmtravlr commented on src/main/java/com/tmtravlr/additions/gui/type/card/GuiAdditionCardPotionType.java:
```diff
@@ -0,0 +1,108 @@
+
+        this.addition = addition;
+
+        this.textCount = TextFormatting.GRAY + I18n.format("gui.view.additionType.count", TextFormatting.RESET + String.valueOf(this.addition.getEffectCount()) + TextFormatting.GRAY);
```
Hmm, maybe this should have a new translated string that says `Potion Effects: <number>` instead of `Count: <number>`. Might make it more obvious what it is a count for.

Tmtravlr commented on src/main/java/com/tmtravlr/additions/gui/view/edit/GuiEditPotionType.java:
```diff
@@ -0,0 +1,195 @@
+        String baseName = CommonGuiUtils.requireStringField(this.baseNameInput, this, "gui.edit.potionType.problem.title", new TextComponentTranslation("gui.edit.potionType.problem.noBase"));
+        if (baseName == null) return;
+
+        String base = TextFormatting.getTextWithoutFormattingCodes(baseName), formats=baseName.substring(0, baseName.indexOf(base));
```
Code style nitpick: Don't set two strings on the same line, it makes the code harder to read

Tmtravlr commented on src/main/java/com/tmtravlr/additions/gui/view/edit/GuiEditPotionType.java:
```diff
@@ -0,0 +1,195 @@
+        List<GuiComponentPotionEffectInput> effectInputs = this.potionEffectListInput.getComponents();
+        for (int i = 0; i < effectInputs.size(); i++) {
+            effects[i] = effectInputs.get(i).getPotionEffect();
+        }
```
You can use a stream for this!

Edit: Added a step to remove nulls
```suggestion
        PotionEffect[] effects = effectInputs.getComponents().stream()
                .map(GuiComponentPotionEffectInput::getPotionEffect)
                .filter(effect -> effect != null)
                .collect(Collectors.toList())
                .toArray(new PotionEffect[0]);
```

Tmtravlr commented on src/main/java/com/tmtravlr/additions/gui/view/edit/GuiEditPotionType.java:
```diff
@@ -0,0 +1,195 @@
+    }
+
+    public void copyFromOther() {
+        this.potionEffectListInput.removeAllComponents();
+        this.copyFrom.getEffects().forEach(toAdd -> {
+            GuiComponentPotionEffectInput input = this.potionEffectListInput.createBlankComponent();
+            input.setDefaultPotionEffect(toAdd);
+            this.potionEffectListInput.addDefaultComponent(input);
+        });
```
Need to add the names and color

Tmtravlr commented on src/main/java/com/tmtravlr/additions/util/GeneralUtils.java:
```diff
@@ -0,0 +1,48 @@
+     * @param <S> a generic type to support all the serializers
+     * @return a completed Gson object
+     */
+    public static <S> Gson newGson(Class clazz, S serializer) {
```
I don't think you actually used this one, did you?

Tmtravlr commented on src/main/java/com/tmtravlr/additions/gui/view/edit/GuiEditPotionType.java:
```diff
@@ -0,0 +1,195 @@
+            effects[i] = effectInputs.get(i).getPotionEffect();
+        }
+
+        PotionTypeAdded type = new PotionTypeAdded(name, baseName, splashName, lingeringName, arrowName, GeneralUtils.removeNulls(effects)).setColor(potionColor);
```
Actually you can add the remove nulls part to the stream too. I'll add it to that comment.

Tmtravlr commented on src/main/java/com/tmtravlr/additions/util/GeneralUtils.java:
```diff
@@ -0,0 +1,48 @@
+     * @param <A> the type of array
+     * @return input without any elements of null
+     */
+    public static <A> A[] removeNulls(A[] input) {
```
You can actually just remove nulls in the stream I mentioned above, so this method isn't needed

sschr15 committed:
Removed untranslated things

sschr15 committed:
Apply most changes requested by @Tmtravlr

Signed-off-by: sschr15 <steven.sschr15@hotmail.com>

sschr15 committed:
Switch LootOverhaul to using CurseMaven instead of a file in the libs.

sschr15 committed:
cleaned up too much stuffs

sschr15 committed:
get it to build, hopefully

sschr15 committed:
add gradle wrapper (gradlew, gradlew.bat, gradle/wrapper/*)
